File: app/src/main/java/es/etologic/mahjongscoring2/app/model/GameItemDiffUtilCallback.java
package es.etologic.mahjongscoring2.app.model;

import android.support.v7.util.DiffUtil;

import java.util.List;

import es.etologic.mahjongscoring2.app.utils.DateTimeUtils;
import es.etologic.mahjongscoring2.domain.model.BestHand;
import es.etologic.mahjongscoring2.domain.model.GameWithRounds;
import es.etologic.mahjongscoring2.domain.model.Round;

public class GameItemDiffUtilCallback extends DiffUtil.Callback {

    private static final int NUM_GAME_PLAYERS = 4;
    private List<GameWithRounds> oldList;
    private List<GameWithRounds> newList;

    public GameItemDiffUtilCallback(List<GameWithRounds> newList, List<GameWithRounds> oldList) {
        this.newList = newList;
        this.oldList = oldList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getGame().getGameId() ==
                newList.get(newItemPosition).getGame().getGameId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        GameWithRounds newGame = newList.get(newItemPosition);
        GameWithRounds oldGame = oldList.get(oldItemPosition);
        return oldGame.getGame().getGameId() == newGame.getGame().getGameId() &&
                oldGame.getGame().getNameP1().equals(newGame.getGame().getNameP1()) &&
                oldGame.getGame().getNameP2().equals(newGame.getGame().getNameP2()) &&
                oldGame.getGame().getNameP3().equals(newGame.getGame().getNameP3()) &&
                oldGame.getGame().getNameP4().equals(newGame.getGame().getNameP4()) &&
                DateTimeUtils.areEqual(oldGame.getGame().getStartDate(), newGame.getGame().getStartDate()) &&
                arePlayersTotalsPointsEquals(oldGame.getPlayersTotalPointsString(),
                        newGame.getPlayersTotalPointsString()) &&
                areBestHandsEqual(oldGame.getBestHand(), newGame.getBestHand())
                && Round.areEqual(oldGame.getRounds(), newGame.getRounds());
    }

    private boolean arePlayersTotalsPointsEquals(String[] oldPlayersTotalPoints,
                                                 String[] newPlayersTotalPoints) {
        for (int i = 0; i < NUM_GAME_PLAYERS; i++) {
            if (!oldPlayersTotalPoints[i].equals(newPlayersTotalPoints[i])) {
                return false;
            }
        }
        return true;
    }

    private boolean areBestHandsEqual(BestHand oldBestHand, BestHand newBestHand) {
        return oldBestHand.getPlayerName().equals(newBestHand.getPlayerName()) &&
                oldBestHand.getHandValue() == newBestHand.getHandValue();
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/model/GameItemDiffUtilCallback.kt
package es.etologic.mahjongscoring2.app.model

import androidx.recyclerview.widget.DiffUtil
import es.etologic.mahjongscoring2.app.utils.DateTimeUtils
import es.etologic.mahjongscoring2.domain.model.BestHand
import es.etologic.mahjongscoring2.domain.model.GameWithRounds
import es.etologic.mahjongscoring2.domain.model.Round

class GameItemDiffUtilCallback(private val newList: List<GameWithRounds>, private val oldList: List<GameWithRounds>) : DiffUtil.Callback() {
    
    companion object {
        private const val NUM_GAME_PLAYERS = 4
    }
    
    override fun getOldListSize(): Int {
        return oldList.size
    }
    
    override fun getNewListSize(): Int {
        return newList.size
    }
    
    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].game.gameId == newList[newItemPosition].game.gameId
    }
    
    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val newGame = newList[newItemPosition]
        val oldGame = oldList[oldItemPosition]
        return (oldGame.game.gameId == newGame.game.gameId &&
            oldGame.game.nameP1 == newGame.game.nameP1 &&
            oldGame.game.nameP2 == newGame.game.nameP2 &&
            oldGame.game.nameP3 == newGame.game.nameP3 &&
            oldGame.game.nameP4 == newGame.game.nameP4 &&
            DateTimeUtils.areEqual(oldGame.game.startDate, newGame.game.startDate) &&
            arePlayersTotalsPointsEquals(
                oldGame.getPlayersTotalPointsString(),
                newGame.getPlayersTotalPointsString()
            ) &&
            areBestHandsEqual(oldGame.getBestHand(), newGame.getBestHand())
            && Round.areEqual(oldGame.rounds, newGame.rounds))
    }
    
    private fun arePlayersTotalsPointsEquals(
        oldPlayersTotalPoints: Array<String>,
        newPlayersTotalPoints: Array<String>
    ): Boolean {
        for (i in 0 until NUM_GAME_PLAYERS) {
            if (oldPlayersTotalPoints[i] != newPlayersTotalPoints[i]) {
                return false
            }
        }
        return true
    }
    
    private fun areBestHandsEqual(oldBestHand: BestHand, newBestHand: BestHand): Boolean {
        return oldBestHand.playerName == newBestHand.playerName && oldBestHand.handValue == newBestHand.handValue
    }
}
